feat(stdlib): add encode_charset, decode_charset functions

powerumc opened the pull request:
<!--
  Your PR title must conform to the conventional commit spec:
  https://www.conventionalcommits.org/en/v1.0.0/

  <type>(<scope>)!: <description>

  * `type` = chore, enhancement, feat, fix, docs, revert
  * `!` = OPTIONAL: signals a breaking change
  * `scope` = Optional when `type` is "chore" or "docs"
  * `description` = short description of the change

Examples:

  * enhancement(file source): Add `sort` option to sort discovered files
  * feat(new source): Initial `statsd` source
  * fix(file source): Fix a bug discovering new files
  * chore(external docs): Clarify `batch_size` option
-->

## Summary

Charset encoding and decoding functions were implemented to integrate the charset settings used in various operating systems and systems with vector.

- decode_charset(value, from_charset)
- encode_charset(value, to_charset)


## Change Type

- [ ] Bug fix
- [x] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?

- [ ] Yes
- [x] No

## How did you test this PR?

- [decode_charset unit test](https://github.com/powerumc/vector-vrl-contrib/blob/charset-encode-decode/src/stdlib/decode_charset.rs#L121-L155)
- [encode_charset unit test](https://github.com/powerumc/vector-vrl-contrib/blob/charset-encode-decode/src/stdlib/encode_charset.rs#L117-L151)

## Does this PR include user facing changes?

- [x] Yes. Please add a changelog fragment based on
  our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the "no-changelog" label to this PR.

## Checklist

- [x] Our [CONTRIBUTING.md](https://github.com/vectordotdev/vrl/blob/main/CONTRIBUTING.md) is a good starting place.
- [x] If this PR introduces changes to [LICENSE-3rdparty.csv](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv), please
  run `dd-rust-license-tool write` and commit the changes. More details [here](https://crates.io/crates/dd-rust-license-tool).
- [x] For new VRL functions, please also create a sibling PR in Vector to document the new function.
  - [vectordotdev/vector #21981](https://github.com/vectordotdev/vector/pull/21981) 

PR adding new VRL function: 
- [decode_charset](https://github.com/vectordotdev/vrl/pull/1162/files#diff-a2b29ab68373ed4a5433b9ccae091de7250ac1a3843df8700da2f382667b534c) function
- [encode_charset](https://github.com/vectordotdev/vrl/pull/1162/files#diff-e85c3cc34d764ef2b07194b6b6e6461b165064a3bda596d1c5124afa47aaf643) function

PR adding documentation:
- [1162.feature.md](https://github.com/vectordotdev/vrl/pull/1162/files#diff-98f42faae0e58a0579966173e58ea210c230b547ce8e6f5152c440422be07d76)

<!-- Examples for the above:
  PR adding new VRL function: https://github.com/vectordotdev/vrl/pull/993
  PR adding documentation: https://github.com/vectordotdev/vector/pull/21142
  
  We are working towards improving this workflow.
-->

## References

- Closes: #1155

<!-- Any other issues or PRs relevant to this PR? Feel free to list them here. -->

powerumc committed:
feat(stdlib): add encode_charset, decode_charset functions

Charset encoding and decoding functions were implemented to integrate the charset settings used in various operating systems and systems with vector.

- decode_charset(value, from_charset)
- encode_charset(value, to_charset)

powerumc committed:
test(stdlib): add encode/decode_charset to benches/stdlib.rs

powerumc committed:
docs(stdlib): add changelog.d

pront commented on src/stdlib/decode_charset.rs:
```diff
@@ -0,0 +1,153 @@
+    Ok(Value::Bytes(output.as_bytes().to_vec().into()))
+}
+
+fn error(from: &[u8]) -> ExpressionError {
```
nit: 
```suggestion
fn create_error(from_charset: &[u8]) -> ExpressionError {
```

powerumc replied:
hank you for your interest. 
I have modified it in the commit. d4be5d7a5712e4bf6c89104c35795f8f0e719cfa

pront reviewed (approved):
Thank you @powerumc!

Left a couple of minor comments.

pront commented on src/stdlib/encode_charset.rs:
```diff
@@ -0,0 +1,149 @@
+    Ok(Value::Bytes(output.as_bytes().to_vec().into()))
+}
+
+fn error(from: &[u8]) -> ExpressionError {
```
Nit: This is identical to the one above. Maybe we can move everything under a new directory `src/stdlib/charset`

powerumc committed:
chore(stdlib): Modified the error function and its parameter name

- rename `error` function to `create_error`
- rename `from` to `from_charset` in decode_charset
- rename `from` to `to_charset` in encode_charset

powerumc commented:
@pront 
I submitted a sibling PR in Vector to document the new function. [vectordotdev/vector #21981](https://github.com/vectordotdev/vector/pull/21981)
Thank you for your review. 😊

pront commented:
The licenses check fails, please use `dd-rust-license-tool write && git commit -am "dd-rust-license-tool write"` and push to fix.

powerumc commented:
> The licenses check fails, please use `dd-rust-license-tool write && git commit -am "dd-rust-license-tool write"` and push to fix.

In my case, there is no change in the `dd-rust-license-tool write` and `commit -am "dd-rust-license-tool write"`  command.
the `check-licenses` step in GitHub Actions was successful.

The error occurs in the check-deny step. Is this caused by my code?
https://github.com/vectordotdev/vrl/actions/runs/12160810717/job/34141068366?pr=1162

jszwedko committed:
Merge remote-tracking branch 'origin/main' into charset-encode-decode

jszwedko commented:
I think we just needed `main` merged in to fix the `cargo deny` failure. I merged it here.
